src/api/ledger/pathfind.js

```diff
@@ -1,147 +1,106 @@
+  if (pathfind.source.amounts && pathfind.source.amounts.length > 0) {
+    params.src_currencies = pathfind.source.amounts.map(amount =>
+      _.omit(amount, 'value'));
   }
 
-  function findPath(pathfindParams, _callback) {
-    const request = self.remote.requestRipplePathFind(pathfindParams);
-    request.once('error', _callback);
-    request.once('success', function(pathfindResults) {
-      pathfindResults.source_account = pathfindParams.src_account;
-      pathfindResults.source_currencies = pathfindParams.src_currencies;
-      pathfindResults.destination_amount = pathfindParams.dst_amount;
-      _callback(null, pathfindResults);
-    });
+  remote.requestRipplePathFind(params,
+    composeAsync(_.partial(addParams, params), callback));
+}
 
-    function reconnectRippled() {
-      self.remote.disconnect(function() {
-        self.remote.connect();
-      });
-    }
-    request.timeout(utils.common.server.CONNECTION_TIMEOUT, function() {
-      request.removeAllListeners();
-      reconnectRippled();
-      _callback(new TimeOutError('Path request timeout'));
+function addDirectXrpPath(paths, xrpBalance) {
+  // Add XRP "path" only if the source acct has enough XRP to make the payment
+  const destinationAmount = paths.destination_amount;
+  if ((new BigNumber(xrpBalance)).greaterThanOrEqualTo(destinationAmount)) {
+    paths.alternatives.unshift({
+      paths_canonical: [],
+      paths_computed: [],
+      source_amount: paths.destination_amount
     });
-    request.request();
   }
+  return paths;
+}
 
-  function addDirectXrpPath(pathfindResults, _callback) {
-    // Check if destination_amount is XRP and if destination_account accepts XRP
-    if (typeof pathfindResults.destination_amount.currency === 'string'
-          || pathfindResults.destination_currencies.indexOf('XRP') === -1) {
-      return _callback(null, pathfindResults);
-    }
-    // Check source_account balance
-    self.remote.requestAccountInfo({account: pathfindResults.source_account},
-        function(error, result) {
-      if (error) {
-        return _callback(new Error(
-          'Cannot get account info for source_account. ' + error));
-      }
-      if (!result || !result.account_data || !result.account_data.Balance) {
-        return _callback(new Error('Internal Error. Malformed account info : '
-                                  + JSON.stringify(result)));
-      }
-      // Add XRP "path" only if the source_account has enough money
-      // to execute the payment
-      if (bignum(result.account_data.Balance).greaterThan(
-                                      pathfindResults.destination_amount)) {
-        pathfindResults.alternatives.unshift({
-          paths_canonical: [],
-          paths_computed: [],
-          source_amount: pathfindResults.destination_amount
-        });
-      }
-      _callback(null, pathfindResults);
-    });
-  }
+function isRippledIOUAmount(amount) {
+  // rippled XRP amounts are specified as decimal strings
```

Comment:
Strictly speaking, rippled will never output an XRP amount as anything other than a drops string. It will never output `{"currency" : "XRP", ... }` afaik.